zerver/lib/encryption.py

```diff
@@ -0,0 +1,52 @@
+import base64
+import os
+from typing import Tuple
+
+from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+
+from zerver.models import PushDeviceToken
+
+
+def bytes_to_b64(data: bytes) -> str:
+    """
+    Converts bytes to url-safe base64 encoded string.
+    """
+    return base64.b64encode(data).decode('utf-8')
+
+def b64_to_bytes(data: str) -> bytes:
+    """
+    Converts url-safe base64 encoded string to bytes.
+    """
+    return base64.b64decode(data.encode('utf-8'))
+
+def encrypt_data(key: str, data: str) -> Tuple[str, str]:
+    """
+    Returns a tuple of `(encrypted_data, nonce)`.
+    """
+    # AESGCM() expects key as bytes.
+    aesgcm = AESGCM(b64_to_bytes(key))
+    nonce = os.urandom(12)
+    # AESGCM.encrypt() expects data in bytes and returns ciphertext
+    # bytes with the 16 byte tag appended.
+    encrypted_data = aesgcm.encrypt(nonce, data.encode('utf-8'), None)
+    return bytes_to_b64(encrypted_data), bytes_to_b64(nonce)
+
+def decrypt_data(key: str, nonce: str, encrypted_data: str) -> str:
+    # AESGCM() expects key as bytes.
+    aesgcm = AESGCM(b64_to_bytes(key))
+    # AESGCM.decrypt() expects encrypted data and nonce as bytes.
+    data = aesgcm.decrypt(b64_to_bytes(nonce),
+                          b64_to_bytes(encrypted_data),
+                          None)
+    # Return unicode data.
+    return data.decode('utf-8')
+
+def generate_encryption_key() -> str:
+    while True:
+        key = AESGCM.generate_key(bit_length=128)
+        # generate_key() returns key as bytes. We covert it to base64
+        # and store it as a string in the DB.
+        key_str = bytes_to_b64(key)
+        # Make sure the generated key doesn't collide with an existing key.
+        if PushDeviceToken.objects.filter(notification_encryption_key=key_str).count() == 0:
+            return key_str
```

Comment:
I think that `AESGCM.generate_key` can generate the same keys twice though the chances are very little. We are not generating the encryption key through our code, instead we are relying on `AESGCM.generate_key` to do so.

This change was suggested by @Hypro999 in a [resolved conversation](https://github.com/zulip/zulip/pull/15229#discussion_r449750853) -- so he might want to add something here.

Reply:
I don't really have anything to add here. This was just added as extremely defensive code. We can remove this if we want to, since it involves a DB query for a highly unlikely event.

Reply:
Yes, let's take this out.

In the absence of a bug in how we generate the key (including in the system's random-number generator which we'll be relying on), the probability of a collision is far smaller than the probability of, say, the system going completely wild due to a hardware failure. Even if there are a billion records here, the probability of a collision is less than 1/2^68. The probability that the machine will take a direct hit from a meteor is several orders of magnitude bigger than that.

If OTOH there *is* a bug in the key generation... this would not be an effective way to detect that (and the Zulip app in general isn't the right layer to be attempting to defend against such a thing anyway.)

Reply:
Agreed.